### ISSUE TITLE
Add support for waiting for an exception

### ISSUE THREAD

**Issue by johanhaleby** [awaitility/awaitility#129]

**Comment by olasundell** [comment]
I managed a not-fugly workaround, which could probably be improved:

`        
assertThrows<CacheDiscrepancyException> {
            (await withPollInterval ONE_HUNDRED_MILLISECONDS).atMost(ONE_SECOND).until {
                try {
                    check.check()
                    false
                } catch (e: CacheDiscrepancyException) {
                    true
                }
            }
}
`

**Comment by johanhaleby** [comment]
We _should_ be able to do something like this (once support for it is implemented):

```kotlin
await untilException<CacheDiscrepancyException>()
```

**Comment by johanhaleby** [comment]
@olasundell I've started implementing in a new branch called [wait-for-exceptions](https://github.com/awaitility/awaitility/tree/wait-for-exceptions). It's very much a proof of concept at this point and you can use it like this:

```java
IOException exception = await().atMost(1000, MILLISECONDS).untilThrown(IOException.class, countingCallable);      
```

But I have some questions before I continue. 

1. Do you think `untilThrown` is a good name?
1. Do you think `untilThrown` should also catch subclasses of IOExceptions when used in this way? (and if so should we also have a `untilThrownExactly` that only match if it's _exactly_ an instance of `IOException). My guess is that in most cases you'd also want subclasses of the specified exception. However [ignoreException](https://www.javadoc.io/static/org.awaitility/awaitility/4.0.1/org/awaitility/core/ConditionFactory.html#ignoreException-java.lang.Class-) as its implemented today doesn't allow subclasses. To also include subclasses one should use [ignoreExceptionsInstanceOf](https://www.javadoc.io/static/org.awaitility/awaitility/4.0.1/org/awaitility/core/ConditionFactory.html#ignoreExceptionsInstanceOf-java.lang.Class-). So for consistency maybe we should adhere to the same principle here?
1. Should the exception be define before the condition? I.e. should it be `untilThrown(IOException.class, countingCallable)` or `untilThrown(countingCallable, IOException.class)`. The reason I'm asking is because the other `until` methods in Awaitility takes the condition _first_ and the matcher _second_ (e.g. `await().until(dataRepository:findFirst, equalTo(Data.of("stuff"))) instead of `await().until(equalTo(Data.of("stuff")), dataRepository:findFirst)`). In Kotlin we can create an extension function that will look like this though: `await untilThrow<IOException> { ... }`.

WDYT?

**Comment by olasundell** [comment]
Sorry, I've been away at Kotlinconf so I haven't been able to give your comment the serious answer it demands.
1. I think `untilThrown` is a good name, given the names of the other available `until` methods.
2. I think it should catch subclasses. Exceptions is IMNSHO one of the few real-world cases where inheritance makes perfect sense. I imagine, however, that a less-than-superhuman developer could make a serious mess out of this by ignoring certain exceptions in an inheritance hierarchy, while waiting for `untilThrown` for others within said structure. Yikes. I wish I could be more decisive here. Adhering to the previous principle is important so usage is intuitive, but otoh I might argue that the previous way of doing it isn't quite what I'd recommend. What if `untilThrown` could take a parameter which specifies whether subclasses should be considered or not?
3. The exception should be defined after the condition, for intuitiveness.

**Comment by grofoli** [comment]
@johanhaleby I agree with @olasundell in 1. and 3., so both the method name and the method signature is good like he described: `public <T extends Throwable> T untilThrown(Callable<?> callable, Class<T> type)`. Wanna call the Callable `callable `or `supplier`, maybe?

Regarding 2. I like the suggestion `untilThrown `and `untilThrownExactly`.

**Comment by johanhaleby** [comment]
Thanks a lot for your input, it's really invaluable! I'll definitely take your inputs into account when I continue with the implementation and if you don't mind I'll poke you here if there are more things I run into :)

**Comment by grofoli** [comment]
Yes, sure, feel free to poke me :-)

**Comment by pbabu0192** [comment]
Any update on this?
